src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -711,6 +793,109 @@ - (BOOL)_attributeDataValue:(MTRDeviceDataValueDictionary)observed satisfiesValu
     return YES;
 }
 
+#pragma mark - Handling of waits for attribute values
+
+- (void)waitForAttributeValues:(NSArray<NSDictionary<NSString *, id> *> *)values timeout:(NSNumber * _Nullable)timeout queue:(dispatch_queue_t)queue completion:(void (^)(NSError * _Nullable error))completion
+{
+    // Check that all the values are in fact attribute response-values.
+    for (NSDictionary<NSString *, id> * value in values) {
+        if (!value[MTRAttributePathKey] || !value[MTRDataKey]) {
+            MTR_LOG_ERROR("%@ waitForAttributeValues passed a non-attribute value %@", self, value);
+            dispatch_async(queue, ^{
+                completion([MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT]);
+            });
+            return;
+        }
+    }
+
+    // Check whether we have all these values already.
+    NSMutableArray<MTRAttributeRequestPath *> * requestPaths = [NSMutableArray arrayWithCapacity:values.count];
+    for (NSDictionary<NSString *, id> * value in values) {
+        MTRAttributePath * path = value[MTRAttributePathKey];
+        [requestPaths addObject:[MTRAttributeRequestPath requestPathWithEndpointID:path.endpoint clusterID:path.cluster attributeID:path.attribute]];
+    }
+
+    NSArray<NSDictionary<NSString *, id> *> * currentValues = [self readAttributePaths:requestPaths];
+
+    std::lock_guard lock(_lock);
+    auto * attributeWaiter = [[MTRAttributeValueWaiter alloc] initWithValues:values queue:queue completion:completion];
+
+    for (NSDictionary<NSString *, id> * currentValue in currentValues) {
+        // Pretend as if this got reported, for purposes of the attribute
+        // waiter.
+        [attributeWaiter attributeValue:currentValue[MTRDataKey] reportedForPath:currentValue[MTRAttributePathKey] byDevice:self];
+    }
+
+    if (attributeWaiter.allValuesSatisfied) {
+        MTR_LOG("%@ waitForAttributeValues no need to wait, values already match: %@", self, values);
+        dispatch_async(queue, ^{
+            completion(nil);
+        });
+        return;
+    }
+
+    // Otherwise, wait for the values to arrive or our timeout.
+    [self.attributeValueWaiters addObject:attributeWaiter];
+
+    if (timeout != nil) {
+        mtr_weakify(attributeWaiter);
+        mtr_weakify(self);
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, static_cast<uint64_t>(timeout.doubleValue * NSEC_PER_SEC)), self.queue, ^{
```

Comment:
If timeouts can be long, it would be better to use a dispatch_source that can be cancelled, otherwise the dispatched block will remain in memory until the timeout comes around even if we successfully arrived at the expected state.

Reply:
Added canceling of timers.